test/D2L.Security.OAuth2.UnitTests/Keys/Default/KeyManagementServiceTests.cs

```diff
@@ -0,0 +1,377 @@
+			);
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( secondKey, creds );
+		}
+
+		[Test]
+		public async Task RefreshAsync_HasVeryNewKeyCached_ReplacesWithLessNewKey() {
+			var firstKey = CreateVeryNewKey();
+
+			await PrimeCacheAsync( firstKey );
+
+			var secondKey = CreateNewKey();
+
+			UsePrivateKeys( firstKey, secondKey );
+
+			await m_kms.RefreshKeyAsync();
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			// We prefer the oldest key in this case because it is the least
+			// likely to cause cache problems
+			AssertCorrectKey( secondKey, creds );
+		}
+
+		[Test]
+		public async Task RefreshAsync_HasVeryOldKeyCached_NoNewKeys_SaysRetrySoon() {
+			var key = CreateVeryOldKey();
+
+			await PrimeCacheAsync( key );
+
+			var actualDt = await m_kms.RefreshKeyAsync();
+
+			Assert.AreEqual( TimeSpan.FromMinutes( 1 ), actualDt );
+		}
+
+		[Test]
+		public async Task RefreshAsync_NoKeys_SaysRetryVerySoon() {
+			UsePrivateKeys();
+
+			var actualDt = await m_kms.RefreshKeyAsync();
+
+			Assert.AreEqual( TimeSpan.FromSeconds( 10 ), actualDt );
+
+			// Double check that we didn't cache anything
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			Assert.IsNull( creds );
+		}
+
+		[Test]
+		public async Task GenerateNewKeyIfNeededAsync_HasActiveKey_DoesNothing() {
+			var key = CreateActiveKey();
+
+			UsePrivateKeys( key );
+
+			await m_kms.RefreshKeyAsync();
+
+			// If it were to generate a new key it would have to save it which
+			// would trigger our strict mocks.
+		}
+
+		[Test]
+		public async Task GenerateNewKeyIfNeededAsync_HasNewKey_DoesNothing() {
+			var key = CreateNewKey();
+
+			UsePrivateKeys( key );
+
+			// Although we'd prefer to use an older key, generating a new one
+			// couldn't help: the reason NotBefore exists is to give time for
+			// caches to expire, and generating a new key (even with an
+			// artificially in-the-past NotBefore) can't change that.
+
+			await m_kms.GenerateNewKeyIfNeededAsync();
+
+			// Again, relying on strict mocks.
+		}
+
+		[Test]
+		public async Task GenerateNewKeyIfNeededAsync_NoKey_GeneratesOne() {
+			UsePrivateKeys();
+
+			ExpectNewKey();
+
+			await m_kms.GenerateNewKeyIfNeededAsync();
+
+			VerifyNewKeyWasSaved();
+		}
+
+		[Test]
+		public async Task GenerateNewKeyIfNeededAsync_OldKey_GeneratesOne() {
+			var key = CreateOldKey();
+
+			UsePrivateKeys( key );
+
+			ExpectNewKey();
+
+			await m_kms.GenerateNewKeyIfNeededAsync();
+
+			VerifyNewKeyWasSaved();
+		}
+
+		private void ExpectNewKey() {
+			m_publicKeys.Setup( pub => pub.SaveAsync( It.IsAny<Guid>(), It.IsAny<JsonWebKey>() ) )
+				.Returns( Task.CompletedTask )
+				.Verifiable();
+
+			m_privateKeys.Setup( priv => priv.SaveAsync( It.IsAny<PrivateKeyData>() ) )
+				.Returns( Task.CompletedTask )
+				.Verifiable();
+		}
+
+		private void VerifyNewKeyWasSaved() {
+			m_publicKeys.VerifyAll();
+			m_privateKeys.VerifyAll();

```

Comment:
In case you don't remember verifiable mocks (they suck), if those `Setup`s above aren't hit at least once this will throw (confirmed manually.) If they are called multiple times it will pass though.

I'm kind of thinking that in an ideal world this library would have one interface for callers to implement (for key storage) and things would look a bit different.

Trying to strike a balancer here between making tests that test something and not getting too into the implementation details. 🤷